### PR TITLE
Feature/search with classification terms

### PR DESCRIPTION
**Pull request by PenghaiZhang**
#1306 

##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [ ] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

The search query now should be a combination of the query put in the SearchBar and selected Classification terms, and their relationship is `AND`. The relationship between each Classification terms is `OR`.

For example, if the query is `Java`, raw mode is off,  and selected Classification terms are `Hobart` and `Sydney`, the final query sent to server should be `Java* AND (Hobart OR Syndey)`.

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/js/tsrc/modules/SearchModule.ts [hunk 1]
```diff
@@ -106,6 +106,40 @@ export const searchItems = ({
...
+
+  const or = " OR ";
+  const and = " AND ";
+  // Each term should be concatenated by 'OR'.
+  const terms: string =
+    "(" +
+    classificationTerms.reduce((previousTerm, nextTerm) =>
+      previousTerm.concat(or, nextTerm)
+    ) +
+    ")";
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/modules/SearchModule.ts:140]:
```suggestion
  const terms: string =
    `(${classificationTerms.join(or)})`;
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/modules/SearchModule.ts [hunk 2]
```diff
@@ -106,6 +106,40 @@ export const searchItems = ({
...
+  return textQuery ? textQuery.concat(and, terms) : terms;
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/modules/SearchModule.ts:141]:
```suggestion
  return textQuery ? (textQuery+ and + terms) : terms;
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/modules/SearchModule.ts [hunk 3]
```diff
@@ -147,6 +181,10 @@ export interface SearchOptions {
    * Filter search results to only include items with the specified statuses.
    */
   status?: OEQ.Common.ItemStatus[];
+  /**
+   * A list of classification terms.
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/modules/SearchModule.ts:185]:
```suggestion
   * A list of classification terms. (Typically a subset of those generated by Search Facets.)
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/modules/SearchModule.ts [hunk 4]
```diff
@@ -106,6 +106,40 @@ export const searchItems = ({
...
+ * It firstly formats the query based on whether raw mode is on or off.
+ * If there are no Classifications terms selected, it returns the formatted query.
+ * Otherwise, it consolidates all terms into one Lucene query by OR, and combines the two queries
+ * by AND, and return the combined one.
+ *
+ * @param query The query put in the SearchBar.
+ * @param rawMode Whether raw mode is on or off.
+ * @param classificationTerms A list of selected Classification terms.
+ */
+export const processQuery = (
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/modules/SearchModule.ts:120]: Do we need to use this externally? Can `export` be removed?

**Review by edalex-ian — CHANGES_REQUESTED** [review]
Great @PenghaiZhang , this will do what we need. :+1: 

But could we please change a few things (including some re-ordering)? As part of that, could you maybe also move `SearchOptions` from the bottom of the file up to be with the other `interface`s? Makes it consistent with more of our code base with logical grouping.

File: Source/Plugins/Core/com.equella.core/js/tsrc/modules/SearchModule.ts (hunk 1 above)
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/tsrc/modules/SearchModule.ts:140]: This is a lot more concise and readable!

**Review by edalex-ian — APPROVED** [review]
GTG! Merge on dude! :+1:

**Pull request merged**